File: salary2/AddEmployeeTransaction.java
public abstract class AddEmployeeTransaction implements Transaction {
  private int itsEmpId;
  private String itsName;
  private String itsAddress;

  public AddEmployeeTransaction(int empId, String name, String address) {
    itsEmpId = empId;
    itsName = name;
    itsAddress = address;
  }

  public void Execute() {
    PaymentClassification pc = GetClassification();
    PaymentSchedule ps = GetSchedule();
    PaymentMethod pm = new HoldMehotd();
    Employee e = new Employee(itsEmpId, itsName, itsAddress);
    e.SetClassification(pc);
    e.SetSchedule(ps);
    e.SetMethod(pm);
    PayrollDatabase.AddEmployee(itsEmpId, e);
  }
  abstract PaymentSchedule GetSchedule();

  abstract PaymentClassification GetClassification();
}

File: salary/lib/add_employee_transaction.rb
require 'transaction'
require 'employee'

class AddEmployeeTransaction
  include Transaction
  
  attr_accessor :employee_id, :name, :address
  
  def initialize(employee_id, name, address)
    @employee_id = employee_id
    @name = name
    @address = address
  end
  
  def execute
    c = classification()
    s = schedule()
    m = HoldMethod.new
    e = Employee.new(employee_id, name, address)
    e.classification = c
    e.schedule = s
    e.method = m
    PayrollDatabase.instance.add_employee(employee_id, e)
  end
  
  def classification
    raise 'must implement in subclass'
  end
  
  def schedule
    raise 'must implement in subclass'
  end
  
end

class HoldMethod
end
